Fix compatibility phpv8

TheGivrix opened the pull request:
# Fix #16628 *Compatibility phpv8*
*Fixed errors to get a first validation.*

TheGivrix committed:
Merge pull request #12 from Dolibarr/develop

Update fork

TheGivrix committed:
Merge pull request #13 from Dolibarr/develop

Update

TheGivrix committed:
Merge pull request #14 from Dolibarr/develop

Update fork

TheGivrix committed:
Merge pull request #15 from Dolibarr/develop

Update fork

TheGivrix committed:
Merge pull request #16 from Dolibarr/develop

Set include of js as async

TheGivrix committed:
Fix errors phpv8

eldy commented on htdocs/admin/company.php:
```diff
@@ -666,7 +666,7 @@
 
 print '<tr class="oddeven"><td><label for="SOCIETE_FISCAL_MONTH_START">'.$langs->trans("FiscalMonthStart").'</label></td><td>';
-print $formother->select_month($conf->global->SOCIETE_FISCAL_MONTH_START, 'SOCIETE_FISCAL_MONTH_START', 0, 1, 'maxwidth100').'</td></tr>';
+print (isset($conf->global->SOCIETE_FISCAL_MONTH_START) ? $formother->select_month($conf->global->SOCIETE_FISCAL_MONTH_START, 'SOCIETE_FISCAL_MONTH_START', 0, 1, 'maxwidth100') : '').'</td></tr>';
```
If $conf->global->SOCIETE_FISCAL_MONTH_START is not set, the method $formother->select_month to select the month will not be called at all.

TheGivrix committed:
Update company.php

eldy commented on htdocs/admin/delais.php:
```diff
@@ -289,7 +289,7 @@
 
 
-if ($conf->global->MAIN_DISABLE_METEO != 1) {
+if (isset($conf->global->MAIN_DISABLE_METEO) && $conf->global->MAIN_DISABLE_METEO != 1) {
```
This test is not similar with previous one.
For example, if   $conf->global->MAIN_DISABLE_METEO  is not set, old test was true, new one is false

eldy commented on htdocs/admin/mails.php:
```diff
@@ -48,7 +48,7 @@
 	'__ID__' => 'RecipientIdRecord',
 	//'__EMAIL__' => 'RecipientEMail',				// Done into actions_sendmails
-	'__CHECK_READ__' => (is_object($object) && is_object($object->thirdparty)) ? '<img src="'.DOL_MAIN_URL_ROOT.'/public/emailing/mailing-read.php?tag='.$object->thirdparty->tag.'&securitykey='.urlencode($conf->global->MAILING_EMAIL_UNSUBSCRIBE_KEY).'" width="1" height="1" style="width:1px;height:1px" border="0"/>' : '',
+	'__CHECK_READ__' => (isset($object->thirdparty) && is_object($object) && is_object($object->thirdparty)) ? '<img src="'.DOL_MAIN_URL_ROOT.'/public/emailing/mailing-read.php?tag='.$object->thirdparty->tag.'&securitykey='.urlencode($conf->global->MAILING_EMAIL_UNSUBSCRIBE_KEY).'" width="1" height="1" style="width:1px;height:1px" border="0"/>' : '',
```
If $object i snot an object then object->thirdparty does not exists and will return an error, this test must be after the test is_object($object)

eldy commented on htdocs/admin/mails.php:
```diff
@@ -554,7 +554,7 @@ function initfields()
 
 	// Disable
-	print '<tr class="oddeven"><td>'.$langs->trans("MAIN_DISABLE_ALL_MAILS").'</td><td>'.yn($conf->global->MAIN_DISABLE_ALL_MAILS);
+	print '<tr class="oddeven"><td>'.$langs->trans("MAIN_DISABLE_ALL_MAILS").'</td><td>'.(isset($conf->global->MAIN_DISABLE_ALL_MAILS) ? yn($conf->global->MAIN_DISABLE_ALL_MAILS) : '');
```
Result is not similar

Givriz replied:
What is print after <td> in older version, when $conf->global->MAIN_DISABLE_ALL_MAILS is not set ?

eldy replied:
With old code, if  $conf->global->MAIN_DISABLE_ALL_MAILS  is not set, result is
print ...<td>.yn(unset value);

And the function yn is executed. With your change, the function is not executed.

eldy commented on htdocs/admin/mails.php:
```diff
@@ -737,26 +737,28 @@ function initfields()
-				$emailsenderprofile = new EmailSenderProfile($db);
-				$emailsenderprofile->fetch($id);
-				print $emailsenderprofile->label.' '.dol_escape_htmltag('<'.$emailsenderprofile->email.'>');
+		if (isset($conf->global->MAIN_MAIL_DEFAULT_FROMTYPE)) {
```
This test is not similar, It will never reach the else.

Givriz replied:
How can I preg_replace if $conf->global->MAIN_MAIL_DEFAULT_FROMTYPE isn't set ?
I guess I have to initialise it, but how sould I do ?

eldy replied:
When a value is not set and PHP need to process it, PHP autoconvert it into 0 or ''
So here, if we want to change without knowing what the code do, we should replace the 
`if ($conf->global->MAIN_MAIL_DEFAULT_FROMTYPE === 'robot') {`
into
`if (isset($conf->global->MAIN_MAIL_DEFAULT_FROMTYPE) && $conf->global->MAIN_MAIL_DEFAULT_FROMTYPE === 'robot') {`

Givriz replied:
Ok, got it for if condition. But then, we will reach the else. How preg_replace() will manage an unset value ?
Sould we do this :
preg_replace('/senderprofile_/', '', isset($conf->global->MAIN_MAIL_DEFAULT_FROMTYPE) ? $conf->global->MAIN_MAIL_DEFAULT_FROMTYPE : null);

eldy commented on htdocs/admin/mails.php:
```diff
@@ -776,7 +778,7 @@ function initfields()
 
 		//Add user to select destinaries list
-		print '<tr class="oddeven"><td>'.$langs->trans("MAIN_MAIL_ENABLED_USER_DEST_SELECT").'</td><td>'.yn($conf->global->MAIN_MAIL_ENABLED_USER_DEST_SELECT).'</td></tr>';
+		print '<tr class="oddeven"><td>'.$langs->trans("MAIN_MAIL_ENABLED_USER_DEST_SELECT").'</td><td>'.(isset($conf->global->MAIN_MAIL_ENABLED_USER_DEST_SELECT) ? yn($conf->global->MAIN_MAIL_ENABLED_USER_DEST_SELECT) : '').'</td></tr>';
```
Result is not similar, the yn is not called in some cases.

eldy commented on htdocs/core/boxes/box_dolibarr_state_board.php:
```diff
@@ -66,7 +66,7 @@ public function __construct($db, $param = '')
 
-		$this->hidden = !($user->rights->societe->lire && empty($user->socid));
+		$this->hidden = !(isset($user->rights->societe->lire) && empty($user->socid));
 	}
```
Result is not similar if $user->rights->societe->lire = 0

eldy commented on htdocs/core/lib/admin.lib.php:
```diff
@@ -1885,14 +1885,14 @@ function email_admin_prepare_head()
 		$h++;
 
-		if ($conf->mailing->enabled) {
+		if (isset($conf->mailing->enabled)) {
```
Result is not similar if $conf->mailing->enabled = 0

eldy commented on htdocs/core/modules/modDataPolicy.class.php:
```diff
@@ -132,7 +132,7 @@ public function __construct($db)
 		);
 
-		$country = explode(":", $conf->global->MAIN_INFO_SOCIETE_COUNTRY);
+		if (isset($conf->global->MAIN_INFO_SOCIETE_COUNTRY)) $country = explode(":", $conf->global->MAIN_INFO_SOCIETE_COUNTRY);
```
if $conf->global->MAIN_INFO_SOCIETE_COUNTRY is not defined, the variable $country is undefined when it should be defined to empty array.

TheGivrix committed:
Update delais.php

TheGivrix committed:
Using empty instead of isset

I used empty instead of isset, and I fixed some problems of my previous pull request

TheGivrix committed:
Update modDataPolicy.class.php

eldy commented on htdocs/admin/mails.php:
```diff
@@ -737,14 +737,14 @@ function initfields()
 		print '<tr class="oddeven"><td>'.$langs->trans('MAIN_MAIL_DEFAULT_FROMTYPE').'</td>';
 		print '<td>';
-		if ($conf->global->MAIN_MAIL_DEFAULT_FROMTYPE === 'robot') {
+		if (!empty($conf->global->MAIN_MAIL_DEFAULT_FROMTYPE) === 'robot') {
```
!empty($conf->global->MAIN_MAIL_DEFAULT_FROMTYPE) is a boolean. Comparing with 'robot' will always be false.

TheGivrix committed:
Update mails.php

TheGivrix committed:
Update modDataPolicy.class.php

There was useless condition
